File: ios/runanywhere-sdks-latest/sdk/runanywhere-flutter/packages/runanywhere_onnx/lib/onnx.dart
/// ONNX Runtime backend for RunAnywhere Flutter SDK.
///
/// This module provides STT, TTS, and VAD capabilities via ONNX Runtime.
/// It is a **thin wrapper** that registers the C++ backend with the service registry.
///
/// ## Architecture (matches Swift/Kotlin)
///
/// The C++ backend (RABackendONNX) handles all business logic:
/// - Service provider registration
/// - Model loading and inference for STT/TTS/VAD
/// - Streaming transcription
///
/// This Dart module just:
/// 1. Calls `rac_backend_onnx_register()` to register the backend
/// 2. The core SDK handles all operations via component APIs
///
/// ## Quick Start
///
/// ```dart
/// import 'package:runanywhere_onnx/runanywhere_onnx.dart';
///
/// // Register the module (matches Swift: ONNX.register())
/// await Onnx.register();
///
/// // Register models through RunAnywhere.models.
/// // The commons registry/router owns framework selection and routing.
/// ```
library;

import 'dart:async';

import 'package:runanywhere/foundation/logging/sdk_logger.dart';
import 'package:runanywhere/native/types/basic_types.dart';
import 'package:runanywhere_onnx/native/onnx_bindings.dart';

/// ONNX Runtime module for STT, TTS, and VAD services.
///
/// Provides speech-to-text, text-to-speech, and voice activity detection
/// capabilities using ONNX Runtime with models like Whisper, Piper, and Silero.
///
/// Matches Swift `ONNX` enum from ONNXRuntime/ONNX.swift.
class Onnx {
  Onnx._();

  // ============================================================================
  // Module Info (matches Swift exactly)
  // ============================================================================

  /// Current version of the ONNX Runtime module
  static const String version = '2.0.0';

  /// ONNX Runtime library version (underlying C library)
  static const String onnxRuntimeVersion = '1.24.3';

  // ============================================================================
  // Registration State
  // ============================================================================

  static bool _isRegistered = false;
  static bool _isSherpaRegistered = false;
  static OnnxBindings? _bindings;
  static final _logger = SDKLogger('Onnx');

  // ============================================================================
  // Registration (matches Swift ONNX.register() exactly)
  // ============================================================================

  /// Register ONNX backend with the C++ service registry.
  ///
  /// This calls `rac_backend_onnx_register()` to register the generic ONNX
  /// backend (embeddings) and then registers the Sherpa-ONNX
  /// engine plugin so STT (Whisper/Zipformer/Paraformer), TTS (Piper/VITS),
  /// and Sherpa-backed VAD models with `framework == .sherpa` can resolve
  /// through the unified plugin router. The two plugins are peers: "onnx"
  /// owns embeddings, "sherpa" owns speech primitives. Mirrors Swift
  /// `ONNX.register()` (ONNXRuntime/ONNX.swift).
  ///
  /// Safe to call multiple times - subsequent calls are no-ops.
  static Future<void> register({int priority = 100}) async {
    if (_isRegistered) {
      _logger.debug('ONNX already registered');
      return;
    }

    // Check native library availability
    if (!isAvailable) {
      _logger.error('ONNX native library not available');
      return;
    }

    _logger.info('Registering ONNX backend with C++ registry...');

    try {
      _bindings = OnnxBindings();
      final result = _bindings!.register();

      // RAC_SUCCESS = 0, RAC_ERROR_MODULE_ALREADY_REGISTERED = specific code
      if (result != RacResultCode.success &&
          result != RacResultCode.errorModuleAlreadyRegistered) {
        _logger.warning('C++ backend registration returned: $result');
        return;
      }

      _isRegistered = true;
      _logger.info('ONNX backend registered successfully (embeddings + plugin)');

      _registerSherpa();
    } catch (e) {
      _logger.error('OnnxBindings not available: $e');
    }
  }

  /// Register the Sherpa-ONNX engine plugin so the commons plugin router can
  /// resolve `framework == .sherpa` models. Mirrors Swift
  /// `ONNX.registerSherpaPlugin()`.
  static void _registerSherpa() {
    if (_isSherpaRegistered) return;
    final bindings = _bindings;
    if (bindings == null) return;

    final result = bindings.registerSherpa();
    if (result == RacResultCode.success ||
        result == RacResultCode.errorModuleAlreadyRegistered) {
      _isSherpaRegistered = true;
      _logger.info(
        'Sherpa engine plugin registered (STT + TTS + VAD via Sherpa-ONNX)',
      );
    } else if (result == RacResultCode.errorNotSupported) {
      _logger.warning(
        'Sherpa engine plugin entry not exported by this build '
        '— Sherpa STT/TTS/VAD will not route',
      );
    } else {
      _logger.error('Sherpa plugin registration failed: $result');
    }
  }

  /// Unregister the ONNX backend from C++ registry.
  static void unregister() {
    if (_isSherpaRegistered) {
      _bindings?.unregisterSherpa();
      _isSherpaRegistered = false;
      _logger.info('Sherpa engine plugin unregistered');
    }

    if (!_isRegistered) return;

    _bindings?.unregister();
    _isRegistered = false;
    _logger.info('ONNX backend unregistered');
  }

  /// Check if the native backend is available on this platform.
  ///
  /// On iOS: Checks DynamicLibrary.process() for statically linked symbols
  /// On Android: Checks if librac_backend_onnx_jni.so can be loaded
  static bool get isAvailable => OnnxBindings.checkAvailability();

  // ============================================================================
  // Cleanup
  // ============================================================================

  /// Dispose of resources
  static void dispose() {
    _bindings = null;
    _isRegistered = false;
    _isSherpaRegistered = false;
    _logger.info('ONNX disposed');
  }

  // ============================================================================
  // Auto-Registration (matches Swift exactly)
  // ============================================================================

  /// Enable auto-registration for this module.
  /// Call this function to trigger C++ backend registration.
  static void autoRegister() {
    unawaited(register());
  }
}

File: ios/runanywhere-sdks-latest/sdk/runanywhere-swift/Sources/ONNXRuntime/ONNX.swift
//
//  ONNX.swift
//  ONNXRuntime Module
//
//  Unified ONNX module - thin wrapper that calls C++ backend registration.
//  This replaces both ONNXRuntime.swift and ONNXServiceProvider.swift.
//

import CRACommons
import ONNXBackend
import RunAnywhere

// MARK: - ONNX Module

/// ONNX Runtime module for STT, TTS, and VAD services.
///
/// Provides speech-to-text, text-to-speech, and voice activity detection
/// capabilities using ONNX Runtime with models like Whisper, Piper, and Silero.
///
/// ## Registration
///
/// ```swift
/// import ONNXRuntime
///
/// // Register the backend (done automatically if auto-registration is enabled)
/// try ONNX.register()
/// ```
///
/// ## Usage
///
/// Services are accessed through the main SDK APIs - the C++ backend handles
/// service creation and lifecycle internally:
///
/// ```swift
/// // STT via public API
/// let text = try await RunAnywhere.transcribe(audioData)
///
/// // TTS via public API
/// try await RunAnywhere.speak("Hello")
/// ```
public enum ONNX {
    private static let logger = SDKLogger(category: "ONNX")

    // MARK: - Module Info

    /// Current version of the ONNX Runtime module
    public static let version = "2.0.0"

    /// ONNX Runtime library version (underlying C library)
    public static let onnxRuntimeVersion = RAVersions.onnxRuntimeIOS

    // MARK: - Registration State

    @MainActor private static var isRegistered = false
    @MainActor private static var isSherpaRegistered = false

    // MARK: - Registration

    /// Register ONNX backend with the C++ service registry.
    ///
    /// This calls `rac_backend_onnx_register()` to register the generic ONNX
    /// module (embeddings + Silero VAD), and then registers the Sherpa-ONNX
    /// engine plugin so STT (Whisper/Zipformer) and TTS (Piper/VITS) models
    /// where `framework == .sherpa` can resolve through the unified plugin
    /// router. The two plugins are peers: "onnx" owns embeddings, "sherpa"
    /// owns speech primitives.
    ///
    /// Safe to call multiple times - subsequent calls are no-ops.
    ///
    /// - Parameter priority: Ignored (C++ uses its own priority system)
    @MainActor
    public static func register(priority _: Int = 100) {
        guard !isRegistered else {
            logger.debug("ONNX already registered, returning")
            return
        }

        logger.info("Registering ONNX backend with C++ registry...")

        let result = rac_backend_onnx_register()

        // RAC_ERROR_MODULE_ALREADY_REGISTERED is OK
        if result != RAC_SUCCESS && result != RAC_ERROR_MODULE_ALREADY_REGISTERED {
            let errorMsg = String(cString: rac_error_message(result))
            logger.error("ONNX registration failed: \(errorMsg)")
            // Don't throw - registration failure shouldn't crash the app
            return
        }

        isRegistered = true
        logger.info("ONNX backend registered successfully (embeddings + plugin)")

        // Register the Sherpa-ONNX engine plugin for STT / TTS / VAD. The
        // `rac_backend_sherpa_register` wrapper is not exported from the
        // shipped RABackendSherpa.xcframework, so we call the plugin
        // registry directly with the entry symbol that IS exported.
        registerSherpaPlugin()
    }

    /// Register the Sherpa-ONNX unified-ABI engine plugin so the commons
    /// plugin router can resolve `framework == .sherpa` models (STT:
    /// Whisper / Zipformer / Paraformer; TTS: Piper / VITS; VAD: Silero).
    @MainActor
    private static func registerSherpaPlugin() {
        guard !isSherpaRegistered else { return }

        guard let vtable = rac_plugin_entry_sherpa() else {
            // warning level so this surfaces even when the logger is still
            // on its production default (.warning) during early-boot backend
            // registration (before Logging.shared.applyEnvironmentConfiguration
            // drops the filter to .debug in Phase 1).
            logger.warning("Sherpa plugin entry returned null — Sherpa STT/TTS/VAD will not route")
            return
        }

        let registerResult = vtable.withMemoryRebound(
            to: rac_engine_vtable_t.self, capacity: 1
        ) { typedPointer -> rac_result_t in
            return rac_plugin_register(typedPointer)
        }

        if registerResult == RAC_SUCCESS ||
           registerResult == RAC_ERROR_MODULE_ALREADY_REGISTERED {
            isSherpaRegistered = true
            logger.warning("Sherpa engine plugin registered (STT + TTS + VAD via Sherpa-ONNX)")
        } else {
            let errorMsg = String(cString: rac_error_message(registerResult))
            // The shipped RABackendSherpa.xcframework is known to be compiled
            // with `RAC_SHERPA_ROUTABLE=0` (the speech-ops #define was OFF at
            // build time), so `sherpa_capability_check()` returns
            // RAC_ERROR_BACKEND_UNAVAILABLE which the plugin registry maps to
            // RAC_ERROR_CAPABILITY_UNSUPPORTED here. Rebuilding the xcframework
            // with RAC_BACKEND_SHERPA=ON + SHERPA_ONNX_AVAILABLE=1 unblocks
            // this; until then STT / TTS via `framework == .sherpa` will fail
            // at loadModel() time with the same message.
            logger.error("Sherpa plugin registration failed: \(errorMsg)")
        }
    }

    /// Unregister the ONNX backend from C++ registry.
    ///
    /// `@MainActor` so the `isRegistered` / `isSherpaRegistered` static flags
    /// stay in the same isolation domain as `register(priority:)` and the
    /// `autoRegister` Task hop. Without this annotation, a teardown call on
    /// a background thread would race the registration path and could leave
    /// the C registry in an inconsistent state (double-unregister or skipped
    /// unregister, see comment record `mlt-003`).
    @MainActor
    public static func unregister() {
        if isSherpaRegistered {
            _ = rac_plugin_unregister("sherpa")
            isSherpaRegistered = false
            logger.info("Sherpa engine plugin unregistered")
        }

        guard isRegistered else { return }

        _ = rac_backend_onnx_unregister()
        isRegistered = false
        logger.info("ONNX backend unregistered")
    }

}

// MARK: - Auto-Registration

extension ONNX {
    /// Enable auto-registration for this module.
    /// Access this property to trigger C++ backend registration.
    public static let autoRegister: Void = {
        Task { @MainActor in
            ONNX.register()
        }
    }()
}
